staging/src/k8s.io/apimachinery/pkg/util/proxy/streamtranslator.go

```diff
@@ -0,0 +1,149 @@
+	"k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/klog/v2"
+)
+
+// StreamTranslatorHandler is a handler which translates WebSocket stream data
+// to SPDY to proxy to kubelet (and ContainerRuntime).
+type StreamTranslatorHandler struct {
+	// Location is the location of the upstream proxy. It is used as the location to Dial on the upstream server
+	// for upgrade requests.
+	Location *url.URL
+	// Transport provides an optional round tripper to use to proxy. If nil, the default proxy transport is used
+	Transport http.RoundTripper
+	// Responder is passed errors that occur while setting up proxying.
+	Responder ErrorResponder
+}
+
+// NewStreamTranslatorHandler creates a new proxy handler. Responder is required for returning
+// errors to the caller.
+func NewStreamTranslatorHandler(location *url.URL, transport http.RoundTripper, responder ErrorResponder) *StreamTranslatorHandler {
+	return &StreamTranslatorHandler{
+		Location:  normalizeLocation(location),
+		Transport: transport,
+		Responder: responder,
+	}
+}
+
+func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Create WebSocket server, including particular streams requested.
+	streamOpts, err := streamOptionsFromRequest(req)
+	if err != nil {
+		klog.Errorf("error creating websocket stream options: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+	websocketStreams, err := webSocketServerStreams(req, w, streamOpts)
+	if err != nil {
+		klog.Errorf("websocket createStreams error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+	defer websocketStreams.conn.Close()
+
+	// Create the SPDY client connection proxied through kubelet.
+	tlsConfig, err := utilnet.TLSClientConfig(h.Transport)
+	if tlsConfig != nil && err == nil {
+		// Manually set NextProtos to "http/1.1", since http/2.0 will NOT upgrade a connection.
+		tlsConfig.NextProtos = []string{"http/1.1"}
+	} else {
+		klog.Infof("Unable to unwrap transport %T to get at TLS config: %v", h.Transport, err)
+	}
+	spdyRoundTripper := spdy.NewRoundTripper(tlsConfig)
+	spdyExecutor, err := spdyproxy.NewSPDYExecutorForTransports(spdyRoundTripper, spdyRoundTripper, "POST", h.Location)
+	if err != nil {
+		klog.Errorf("websocket createStreams error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+
+	// Wire the WebSocket server streams output to the SPDY client input.
+	opts := spdyproxy.StreamOptions{}
+	if streamOpts.stdin {
+		opts.Stdin = websocketStreams.stdinStream
+	}
+	if streamOpts.stdout {
+		opts.Stdout = websocketStreams.stdoutStream
+	}
+	if streamOpts.stderr {
+		opts.Stderr = websocketStreams.stderrStream
+	}
+	if streamOpts.tty {
+		opts.Tty = true
+		opts.TerminalSizeQueue = &translatorSizeQueue{resizeChan: websocketStreams.resizeChan}
+	}
+	// Start the SPDY client with connected streams. Output from the WebSocket server
+	// streams will be forwarded into the SPDY client.
+	err = spdyExecutor.StreamWithContext(context.Background(), opts)
```

Comment:
I think we need to cascade the existing request context instead of using context.Background()

Reply:
Fixed. Cascaded in the passed in request `Context`.